archive/imo/imo2021_q1.lean

```diff
@@ -0,0 +1,264 @@
+import tactic data.real.sqrt
+
+noncomputable theory
+
+open real
+
+/-
+# IMO 2021 Q1
+Let n≥100 be an integer. Ivan writes the numbers n, n+1,..., 2n each on different cards.
+He then shuffles these n+1 cards, and divides them into two piles. Prove that at least one
+of the piles contains two cards such that the sum of their numbers is a perfect square.
+# Solution
+We prove the required claim by constructing a triplet of pairwise unequal
+natural numbers a, b, c ∈ [n, 2n], with each sum of each pair being a perfect square. Then, by
+Pigeonhole principle, at least two numbers in the triplet must lie in the same pile, which
+finishes the proof.
+-/
+lemma lower_bound (n : ℕ) (l : ℤ) (hl : 2 * (l : ℝ) ≥ 2 + sqrt (4 + 2 * n)) :
+  (n : ℤ) + 4 * l ≤ 2 * l * l :=
+begin
+  have h₁ : sqrt (4 + 2 * n) ≤ 2 * l - 2 := by linarith,
+  replace h₁ := (sqrt_le_iff.1 h₁).2,
+  norm_cast at h₁, push_cast at h₁,
+  linarith,
+end
+
+lemma upper_bound (n : ℕ) (l : ℤ) (hl₁ : l ≥ 0) (hl₂ : (l : ℝ) ≤ sqrt (1 + n) - 1):
+2 * l * l + 4 * l ≤ 2 * n :=
+begin
+  have h₁ : (l : ℝ) + 1 ≤ sqrt (1 + n) := by linarith,
+  rw le_sqrt at h₁,
+  ring_exp at h₁,
+  norm_cast at h₁, push_cast at h₁,
+  linarith,
+  repeat {norm_cast, linarith},
```

Comment:
I'd rather these `repeat`s that finish off bonus goals be `all_goals` so it's clear what's going on